lib/pbench/client/__init__.py

```diff
@@ -32,36 +32,42 @@ def __str__(self) -> str:
         return f"API template {self.api} requires {self.cnt} parameters ({self.uri_params})"
 
 
-class Pbench:
+class PbenchServerClient:
     def __init__(self, host: str):
         """
-        Create a Pbench Server client object. This validates that the server is
-        reachable, and responds to a GET on /api/v1/endpoints.
+        Create a Pbench Server client object.
+
+        The connect method should be called to establish a connection and set
+        up the endpoints map before using any other methods.
 
         Args:
-            host    Pbench Server hostname
+            host: Pbench Server hostname
         """
         self.host = host
         self.username = None
         self.auth_token = None
         self.session = None
-        self.connect()
+        self.endpoints: Optional[JSONOBJECT] = None
 
     def _headers(self, user_headers: Optional[Dict[str, str]] = None) -> Dict[str, str]:
         """
-        Create an HTTP request headers dictionary. We always default to
-        "application/json" but each operation allows adding or overriding. If
-        we have an authentication token, then we also assume that we want to
-        use it.
+        Create an HTTP request headers dictionary.
+
+        The connect method can set default HTTP headers which apply to all
+        server calls. This method implicitly adds an authentication token
+        if the client has logged in and also allows the caller to override
+        that or any other default session header. (For example, to change the
+        default "accept" datatype, or to force an invalid authentication token
+        for testing.)
 
         Args:
-            user_headers:   Addition request headers
+            user_headers: Addition request headers
 
         Returns:
             Header dictionary
         """
-        headers = user_headers if user_headers else {}
-        if self.auth_token:
+        headers = CaseInsensitiveDict(user_headers) if user_headers else {}
```

Comment:
Python's happy with the duck typing here, and that just seemed unnecessary and space-inefficient overhead. I don't want to require the callers to work with the buried `requests.structures` type, and I don't think it's necessary. We're assembling whatever keys they provide us into a `CaseInsensitiveDict` to work with; but, again, as there are no keys in `{}` it doesn't matter.

Reply:
You convinced me elsewhere that it doesn't matter.  However, the strong-typing part of me objects to the fact that sometimes this function will return a `CaseInsensitiveDict` and sometimes it will return a standard `Dict` (although, I suppose, if we only claim to return a `Dict`, then returning a subtype of it is not out of bounds...).

Nevertheless, if you don't like
```
        headers = CaseInsensitiveDict(user_headers) if user_headers else CaseInsensitiveDict()
        if self.auth_token and "authorization" not in headers:
            headers["authorization"] = f"Bearer {self.auth_token}"
        return headers
```
how about
```
        headers = CaseInsensitiveDict()
        if self.auth_token:
            headers["authorization"] = f"Bearer {self.auth_token}"
        if user_headers:
            headers.update(user_headers)
        return headers
```
(That's two more lines, but one fewer if-conditional.)

Reply:
> either return quacks like a dict

That's a most excellent turn of phrase!  😁

> if you're going to insist on complicating this

I don't mean to _insist_...I'm just offering opinions and possibilities.  (I think I would use very different words and phrasing if I were insisting.)

The code is acceptable as is (otherwise I would have ticked the "Request Changes" button!), but that doesn't mean that it couldn't be better if you were amenable.

Reply:
> Possibly a variation of your second

Why reverse the order of the two conditions and add the extra logic?  In my suggestion, the `"authorization" not in headers` is handled implicitly by the `update()` call (or so I expected).  For our use cases, the key is usually not present, so the extra check doesn't trigger, wherever it is.

Reply:
> Why reverse the order of the two conditions and add the extra logic? In my suggestion, the `"authorization" not in headers` is handled implicitly by the `update()` call

The logic was that, even if we've logged in and the object has an `auth_token` a caller can still *override* that with a deliberately altered (e.g., invalid) `authorization` header for testing. Yours doesn't allow that. It may or may not be a big deal, but I liked the flexibility and I'd prefer to retain it.

Reply:
> Yours doesn't allow that.

It was meant to...doesn't it?  What I think mine would do is, starting with an empty `CaseInsensitiveDict`, add a `"authorization"` header if we have one and then add the `user_headers` if we have those -- if there's an `"authorization"` header in the `user_headers`, it will supersede any that we added first.  Not so?  It's basically a cascading series of defaults/additions, so it should be maximally flexible...or so I'd hoped.  🙂